### ISSUE TITLE
Goreviewpartner (GRP) , automated review tool, now supports phoenixgo native engine !!

### ISSUE THREAD

**Issue by wonderingabout** [Tencent/PhoenixGo#86]
The support is still incomplete, but it works !!
Congratulations @wodesuck 

GoReviewPartner is a famous sgf analysis tool

it can analyze all moves of a sgf (or only some moves), it can also analyze many sgf automatically in one run

The analysis runs on the background, you can do other things and wait until it is finished, then the generated .rsgf file containing many variations per move for every move can be viewed on the GRP "open .rsgf"

see : https://github.com/pnprog/goreviewpartner/pull/96

and : https://github.com/pnprog/goreviewpartner

![grppgo1](https://user-images.githubusercontent.com/38690718/51589351-f9934380-1ee6-11e9-8f9a-671b15dd3d37.png)
![grppgo2](https://user-images.githubusercontent.com/38690718/51589350-f9934380-1ee6-11e9-8956-20def368f1cf.png)
![grppgo3](https://user-images.githubusercontent.com/38690718/51589348-f9934380-1ee6-11e9-95f4-e68017d7fe97.png)

![prp1](https://user-images.githubusercontent.com/38690718/51589368-03b54200-1ee7-11e9-9685-c050996a4132.png)
![prp2](https://user-images.githubusercontent.com/38690718/51589367-03b54200-1ee7-11e9-8f5c-e419bba0d281.png)
![prp3](https://user-images.githubusercontent.com/38690718/51589366-031cab80-1ee7-11e9-887c-c7e9da5d6f3e.png)

![prp4](https://user-images.githubusercontent.com/38690718/51589380-0fa10400-1ee7-11e9-9b2e-55238840956d.png)
![ppp1](https://user-images.githubusercontent.com/38690718/51589384-129bf480-1ee7-11e9-88fb-083032ef16ba.png)

winrate curve difference 

(part of the difference is winrate black x , white 100-x => yoyo in the graph)

native phoenixgo 3000 sims batch 4 400M tree size 64 max children tensorrt ON

![ppp0c](https://user-images.githubusercontent.com/38690718/51559984-ab008d80-1e83-11e9-8cf4-5dbed3313b47.png)

LZ phoenixgo-orig -v 3200

![lzphoenixgogrp](https://user-images.githubusercontent.com/38690718/51560110-eef39280-1e83-11e9-99bb-3d0d093a4926.png)

**Comment by wonderingabout** [comment]
example of moves analyzed :

**Comment by wonderingabout** [comment]
winrate curve difference 

(part of the difference is winrate black x , white 100-x => yoyo in the graph)

native phoenixgo

![ppp0c](https://user-images.githubusercontent.com/38690718/51559984-ab008d80-1e83-11e9-8cf4-5dbed3313b47.png)

LZ phoenixgo-orig

![lzphoenixgogrp](https://user-images.githubusercontent.com/38690718/51560110-eef39280-1e83-11e9-99bb-3d0d093a4926.png)

**Comment by pnprog** [comment]
Hi @wonderingabout ,

Can you share both RSGF files for me to check? I am not sure why this "yoyo" effect. If it was a consequence of "100 - x", the beginning of the graph (left part) would be different for both engines, but this is not the case. So it must me some other issue.

A "yoyo" effect can be normal, when both players keep failing to see (and play) a game reversing move. But in this case, both engines should have the same result.

**Comment by wonderingabout** [comment]
@pnprog 

here it is :+1: 

http://www.mediafire.com/file/bubs1wi7x86zbk3/game2nativePhoenixGo.rsgf/file
http://www.mediafire.com/file/dtw8taad9hg2qeo/game2LZPG.rsgf/file

**Comment by pnprog** [comment]
Thanks,

I can confirm there are some issues with GRP there. For example, move 27 has two variations for **o17**, which obviously is wrong. The issue is most probably linked with GRP parsing all the log from PhoenixGo.

I will have to fix this before we start comparing both engines. There may be other issues to fix also :)

**Comment by wonderingabout** [comment]
we can use this game as a benchmark when you release some fixes

i'll rerun an analysis and we'll see if its better

**Comment by pnprog** [comment]
Maybe some "edge cases" that appear only when a lot of playout are performed: Maybe the order in which the variations are displayed is not consistent, and GRP got confused or something like this.

I will try to have my computer analyse that move 27 during the whole night, and maybe I will find something inside.

**Comment by wonderingabout** [comment]
take it easy :)

**Comment by alreadydone** [comment]
@godmoves (of [HappyGo](https://github.com/godmoves/HappyGo) expertise) might want to tune in. (See also https://github.com/pnprog/goreviewpartner/pull/96)

**Comment by pnprog** [comment]
@wonderingabout 

> i'll rerun an analysis and we'll see if its better

I just pushed some changes on GRP repository. This should fix the issue we saw above, and also add information about number of playouts (in fact, number of visits). And also use less memory.

Could you have a try again now with that same game, so that we can compare the winrate curves? Don't hesitate to go with bigger tree size (maybe width=6, depth=10?) I am quite confident GRP should handle that properly now.

**Comment by wonderingabout** [comment]
@pnprog 

ok will start it now then :+1: 

but lets just start with 6-6 for starters :)

**Comment by wonderingabout** [comment]
@pnprog 

![analysing](https://user-images.githubusercontent.com/38690718/51784370-f2b73b80-2147-11e9-9608-d4b57793dcf3.png)

gtx1060 power capped at 75W for low temps :+1:

**Comment by wonderingabout** [comment]
also 2 : i'll increase sims to 3200 to have comparison with LZ-PhoenixGo

and also 3 : do i have to use `--logtostderr` in conf ?

**Comment by pnprog** [comment]
yes, logtostderr and verbose 1 are needed.

**Comment by wonderingabout** [comment]
i ask because i saw that you removed `--logtostderr` here : 
(i used it everytime though)

https://github.com/pnprog/goreviewpartner/blob/master/config.ini#L100-L104

since phoenixgo can be a bit hard to configure, can i suggest that we add 2 examples here (1 for tensorrt, and 1 for notensorrt)

```
[PhoenixGo-0]
profile = Example-tensorrt
command = /home/yourusername/PhoenixGo/bazel-bin/mcts/mcts_main
parameters = --gtp --config_path=/home/yourusername/PhoenixGo/etc/mcts_1gpu_grp.conf --logtostderr --v=1
timepermove = 120

profile = Example-notensorrt
command = /home/yourusername/PhoenixGo/bazel-bin/mcts/mcts_main
parameters = --gtp --config_path=/home/yourusername/PhoenixGo/etc/mcts_1gpu_notensorrt_grp.conf --logtostderr --v=1
timepermove = 120
```

i'm sure these examples will be helpful for interested people, because there are a lot of things to pay attention to

i'll also add 2 new config files optimized for grp on phoenixgo github : 
- remove time management to use fixed visits
- remove all other time management options
- mention that it is just an example
- add the debug @alreadydone mentionned, with example of 6x6

also, i would like if you sent a PR to my branch, after all you really helped phoenixgo be better

maybe i can commit with your username and email ?

**Comment by wonderingabout** [comment]
btw time where i am now : 

![grptime](https://user-images.githubusercontent.com/38690718/51784585-48d9ae00-214b-11e9-8e93-2783e2414dc9.png)

**Comment by wonderingabout** [comment]
oh wow i'm so careless lol
forgot to use the new config file, gotta redo the analysis again lol

**Comment by wonderingabout** [comment]
oh no actually i really used the right config file, my bad

**Comment by wonderingabout** [comment]
so here is the result :+1: 

http://www.mediafire.com/file/dvp18509dtj5eru/game2v2.rsgf/file

![graph2](https://user-images.githubusercontent.com/38690718/51785307-14b7ba80-2156-11e9-8472-2bbf6c8d0ef1.png)

still some things to fix :)

**Comment by wonderingabout** [comment]
handicap is a bad idea, zero bots are too strong, max playable is 2 (or 3 for 10k or weaker players, with custom config file)

i dont know about start.sh, but i like to have my own settings :) (especially on windows where you need special syntax)

**Comment by wonderingabout** [comment]
discussion continues here : 

https://github.com/pnprog/goreviewpartner/issues/99

**Comment by pnprog** [comment]
> also, i would like if you sent a PR to my branch, after all you really helped phoenixgo be better

Here it is: https://github.com/pnprog/PhoenixGo/tree/patch-3

In the context of GRP, because time is not critical (there is no risk of loosing a game on time) it is better to let the engine thinks more time if it can help it figure out better the best play. I do something similar for Pachi. In fact, I would go as far as increasing the `time_factor` up to 1.0 (0.3 is the default value).

I don't know how to PR to your branch, github only proposes me to PR to Tencent's master branch at this point.

By the way, I did not see a `mcts_cpu_grp.conf` file :)
And that's a lot of modifications to PhoenixGo's project, that are related to an other project, so they may not be interested in it anyway :)

**Comment by wonderingabout** [comment]
> > also, i would like if you sent a PR to my branch, after all you really helped phoenixgo be better
> 
> Here it is: https://github.com/pnprog/PhoenixGo/tree/patch-3

great :+1: @pnprog 

now you need to go to pull requests of my fork here : 

https://github.com/wonderingabout/PhoenixGo/pulls

then create new pull request 

compare : 

- wonderingabout/PhoenixGo - branch faqv2-bazel-master

with 

- pnprog/PhoenixGo - branch patch-3

**Comment by wonderingabout** [comment]
to answer your questions now, the `grp` files i created have time management all disabled (unlimited time), (all enable time options are set to 0), and timeouts set to 0 ms (infinite time)

i forgot indeed the cpu version (and the cpu profile too, i thought it was included in notensorrt :laughing: ), can you use this opportunity to send it to me ?

**Comment by pnprog** [comment]
I can't find  **wonderingabout/PhoenixGo - branch faqv2-bazel-master** in the menu _Choose a Base Repository_, and I have no clue why. I am really helpless about all those github related things :(

**Comment by wonderingabout** [comment]
yes github is not intuitive to understand at firs, but we all went through that :+1: 

the problem you have is because you forked from Tencent/PhoenixGo, not from wonderingabout/PhoenixGo, so even though you managed to import my branch on your fork of Tencent/PhoenixGo, the automated pull requests point to Tencent/PhoenixGo where the faqv2-bazel branch doesnt exist

to fix that, i think you can write manually the repoowner/project e.g wonderingabout/PhoenixGo and find my branch in pull request

if not, then you can backup all your branches (download zip), then delete Tencent/PhoenixGo (in settings "dangerous zone" : https://help.github.com/articles/deleting-a-repository/ click to see screenshots)

 
then go to wonderingabout/PhoenixGo and fork it here "click fork" : https://github.com/wonderingabout/PhoenixGo

**Comment by wonderingabout** [comment]
i think we can close this issue now,

support questions and research can continue in : https://github.com/pnprog/goreviewpartner/issues/99
